psutil/tests/__init__.py

```diff
@@ -228,35 +227,16 @@ def macos_version():
 
 
 def _get_py_exe():
-    def attempt(exe):
-        try:
-            subprocess.check_call(
-                [exe, "-V"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-        except Exception:
-            return None
-        else:
-            return exe
-
-    if GITHUB_ACTIONS:
-        if PYPY:
-            return which("pypy3") if PY3 else which("pypy")
-        elif FREEBSD:
-            return os.path.realpath(sys.executable)
-        else:
-            return which('python')
-    elif MACOS:
-        exe = \
-            attempt(sys.executable) or \
-            attempt(os.path.realpath(sys.executable)) or \
-            attempt(which("python%s.%s" % sys.version_info[:2])) or \
-            attempt(psutil.Process().exe())
-        if not exe:
-            raise ValueError("can't find python exe real abspath")
-        return exe
-    else:
-        exe = os.path.realpath(sys.executable)
-        assert os.path.exists(exe), exe
-        return exe
```

Comment:
Mmm no, you're removing too much. All this part was gradually built over the years because relying on `sys.executable` alone was not enough. 

I suggest the approach I recommended here:
https://github.com/giampaolo/psutil/pull/2209#discussion_r1134446053
...meaning leaving the existing code as-is, and integrate the new logic on top of it.

Reply:
> All this part was gradually built over the years 

At least the part mentioning GHA can be removed (it either passes or fails CI).

For the rest, it seems to be coming from mostly one commit https://github.com/giampaolo/psutil/commit/d108baf4a27f0f3b115aae5e6109eaf109302bc3, and without an issue or PR discussion around that, it seems mostly a preemptive action to something rather than fixing something, or maybe, it was fixing something at the time which might not be required anymore as the tests would suggest. Do you have any reference other than this commit ?

If keeping this, we might need to reintroduce 2 pythons for test_misc if you're not fine with the usage of sys.executable I'm about to push.

Reply:
> For the rest, it seems to be coming from mostly one commit https://github.com/giampaolo/psutil/commit/d108baf4a27f0f3b115aae5e6109eaf109302bc3, and without an issue or PR discussion around that, it seems mostly a preemptive action to something rather than fixing something, or maybe, it was fixing something at the time which might not be required anymore as the tests would suggest. Do you have any reference other than this commit ?

I did that because on my virtualized OSX box the python exe was a symlink, and that created issues with tests using PYTHON_EXE (most likely they complained about missing path). I haven't turned on VirtualBox in a while, but I think the OSX workaround is still relevant today.

Note: I think you run tests via cibuildwheel and/or virtual env. I use the Makefile on all platforms (including windows via winmake.py), so some workarounds that you see in the code are because of my setup / way of doing things. 

> At least the part mentioning GHA can be removed (it either passes or fails CI).

If it doesn't break tests then sure, but does it? I can't remember why I added logic for PYPI and FREEBSD specifically, but most likely it was because some tests failed on GITHUB. 
The part about PYPY is not used because right now we don't have a PYPY build configured.
Anyway, I'm fine with removing the GITHUB_ACTIONS part as long as it doesn't break tests.

Reply:
Thanks for the background, I will reintroduce that.

> Note: I think you run tests via cibuildwheel and/or virtual env. I use the Makefile on all platforms (including windows via winmake.py), so some workarounds that you see in the code are because of my setup / way of doing things.

Indeed, I always use virtual environment, be it through cibuildwheel, nox, tox, manual setup, ...
I do really think that any development workflow shall not interfere with the system/user installation but that's just my opinion.

Reply:
> I do really think that any development workflow shall not interfere with the system/user installation but that's just my opinion.

I certainly agree. It's an old bad habit I have. FWIW, also the Makefile can be used with venv: 
`make install test PYTHON=/path/to/your/venv`

Reply:
> Anyway, I'm fine with removing the GITHUB_ACTIONS part as long as it doesn't break tests.

I kept it but simplified it.

It seems running linux tests from a virtual environment is failing when just using `exe = os.path.realpath(sys.executable)`. Let's defer this to another PR if needed.

The symptom is `ModuleNotFoundError: No module named 'psutil'`
Using `os.path.realpath(sys.executable)` resolves python outside the virtual env which explains the symptom (also tested with python 3.6).
```
[root@b0438d6ef25a ~]# python3.11 -c 'import sys; import os; print(sys.executable); print(os.path.realpath(sys.executable))'
/usr/local/bin/python3.11
/opt/_internal/cpython-3.11.2/bin/python3.11
[root@b0438d6ef25a ~]# venv-311/bin/python -c 'import sys; import os; print(sys.executable); print(os.path.realpath(sys.executable))'
/root/venv-311/bin/python
/opt/_internal/cpython-3.11.2/bin/python3.11
[root@b0438d6ef25a ~]# virtualenv-311/bin/python -c 'import sys; import os; print(sys.executable); print(os.path.realpath(sys.executable))'
/root/virtualenv-311/bin/python
/opt/_internal/cpython-3.11.2/bin/python3.11
```

Reply:
Wait. You introduced the 2 PYTHONs again (PYTHON_EXE and PYTHON_BASE_EXE). I don't understand why you can't get away with just using one. I'd like to see the problem you have specifically and try debugging it. Can you tell me how I can I reproduce the problem you're experiencing (I'm on linux but I can fire up VMs with other OSes)?

Reply:
OK, I was able to reproduce the errors and fix them with:

```diff
diff --git a/psutil/tests/__init__.py b/psutil/tests/__init__.py
index 61a06e21..28a0c941 100644
--- a/psutil/tests/__init__.py
+++ b/psutil/tests/__init__.py
@@ -240,7 +240,10 @@ def _get_py_exe():
         else:
             return exe

-    if GITHUB_ACTIONS:
+    base_executable = getattr(sys, "_base_executable", None)
+    if WINDOWS and sys.version_info >= (3, 7) and base_executable is not None:
+        return base_executable
+    elif GITHUB_ACTIONS:
         if PYPY:
             return which("pypy3") if PY3 else which("pypy")
         elif FREEBSD:
```

The only one left I have is this, but I'm not sure if it's related with the issue at hand:

```
======================================================================
FAIL: psutil.tests.test_process.TestProcess.test_nice (prio=256)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\user\Desktop\psutil\psutil\tests\test_process.py", line 825, in test_nice
    self.assertEqual(new_prio, prio)
AssertionError: <Priority.HIGH_PRIORITY_CLASS: 128> != 256
----------------------------------------------------------------------
```

What errors do you get if you apply the same patch?

Reply:
And this is how I run tests btw:

```
C:\Users\user\Desktop\psutil>set PYTHON=venv\Scripts\python.exe & make test
```

Reply:
let me try to setup something similar on GHA and share the results on a branch in my fork.

Reply:
This patch is failing (failing to find psutil). It's what I'd expect. https://github.com/mayeut/psutil/actions/runs/4461399430/jobs/7835201234

I'm still trying to adapt the "set PYTHON=venv\Scripts\python.exe & make test" workflow to GHA as I don't have access to a Windows VM otherwise.

Reply:
I tried to reproduce the dev workflow in https://github.com/mayeut/psutil/actions/runs/4461656326 which is also failing with venv.

Reply:
Mmm failed to `"import psutil"` is a different problem than the previous tests that we saw failing, meaning the ones you pasted here (and that were also my failures):
https://github.com/giampaolo/psutil/pull/2209#discussion_r1134563313

It should mean that psutil was installed with one python executable path, and `PYTHON_EXE` var is pointing to another python executable path. At the begging of the build output we can see the value of `PYTHON_EXE`:

```
2023-03-19T16:36:12.5353340Z pyexe:            C:\hostedtoolcache\windows\Python\3.11.2\x64\python.exe
```

That line is being printed by this:
https://github.com/giampaolo/psutil/blob/fb0d5d3325efeb56dd0cf48c0f0d2379baca20d5/psutil/tests/__init__.py#L1178

I suppose one thing you can do is adding more debugging "prints" in there to try to understand the differences between the different paths. E.g.:

```
    info['PYTHON_EXE'] = PYTHON_EXE
    info['sys.executable'] = sys.executable
    info['sys."_base_executable'] =  getattr(sys, "_base_executable", None)
```

Considering in the past I was bitten by symlinks, maybe another thing you can try is `os.path.realpath(sys._base_executable)`.

Reply:
BTW, I added `PYTHON_EXE` / `_get_py_exe()` precisely for this sort of problems. That's what I meant when I said relying on `sys.executable` was not enough. Workarounds are needed depending on the platform / setup / environment.

Reply:
The patch in this PR also works well in the reproduced dev workflow: https://github.com/mayeut/psutil/actions/runs/4465692219/jobs/7843003771

As mentioned in comments & in the code, PYTHON_BASE_EXE / sys._base_executable / or even os.path.realpath(sys._base_executable) won't have psutil installed, by design (it's the global python interpreter, not the venv) & that's why we need 2. In this PR, only `pyrun` reinjects the `__PYVENV_LAUNCHER__` to get psutil importable when using PYTHON_BASE_EXE.
If we don't use the global python interpreter, we can't count processes (or we'd need a specific workaround) or rely on the pid of the launched interpreter to count connections. If we use the global interpreter, we can't import psutil.
My conclusion, getting 2 is the easiest, otherwise, we'd need 1 + an environment variable or rework the failing tests but I feel the 2 interpreters is easier. We can't get away from those issues just by modifying PYTHON_EXE.

Reply:
> My conclusion, getting 2 is the easiest, otherwise, we'd need 1 + an environment variable.

This seems more viable to me. My point is the user of these constants should not think "which one should I use?".  It's an implementation detail that is difficult to understand (at least I still don't fully understand), and which only affects Windows + venv running GITHUB_CI.

I forgot how your previous changes of 1 PYTHON_EXE + env var looked like, but I would try to rehash that approach instead and see how we can arrange it.

Reply:
> I forgot how your previous changes of 1 PYTHON_EXE + env var looked like, but I would try to rehash that approach instead and see how we can arrange it.

Maybe you can do it in a separate PR so that we stop switching between approaches in the same PR.

Reply:
> I forgot how your previous changes of 1 PYTHON_EXE + env var looked like, but I would try to rehash that approach instead and see how we can arrange it.

I've never pushed this one (or even did it in my fork). I will try to open a new PR with this approach this week.

Reply:
I opened 2 other PR with alternative implementations.
The 3 PRs pass the "dev like workflow" on my fork (side-note, it seems the just released pyperf 2.6.0 breaks that workflow, at least the way I implemented it in GHA, so I pinned that to 2.5.0 in my tests of the dev workflow)